### ISSUE TITLE
Compact re-exports for Engines

### ISSUE THREAD

**Issue by trentmwillis** [ember-engines/ember-engines#265]
Currently, when using the `EngineAddon`base class, we automatically re-export dependency's `app` folder contents in the Engine's namespace. This helps make it easy for consumers to use Addons in their Engines. However, the compiled code for this can be quite bloated:

```js
define('ember-chat/components/spanish-greeting', ['exports', 'common-components/components/spanish-greeting'], function (exports, _commonComponentsComponentsSpanishGreeting) {
  'use strict';

  Object.defineProperty(exports, 'default', {
    enumerable: true,
    get: function get() {
      return _commonComponentsComponentsSpanishGreeting['default'];
    }
  });
});
```

For smaller applications, this isn't that big of a deal, but for larger applications with many Engines and many Addons, this duplication can start to add up.

This can be worked around by modifying the resolver to check additional namespaces, but this is sub-optimal as it incurs a runtime penalty for each additional lookup.

What I want to suggest instead is that we leverage the aliasing capabilities of `loader.js` and do the following:

```js
define('ember-chat/components/spanish-greeting', define.alias('common-components/components/spanish-greeting');
```

As far I can tell, this approach is functionally equivalent since the exports should still be the same reference. We could potentially compact this even further and do something to the effect of:

```js
const reexports = [
  ['ember-chat/components/spanish-greeting', 'common-components/components/spanish-greeting'],
  // ...
].forEach(([alias, original]) => define(alias, define.alias(original)));
```

Thoughts?

cc @stefanpenner @nathanhammond @rwjblue

**Comment by nathanhammond** [comment]
Things to think about:

- Does this approach break `instanceOf` checks?
- What do we do in a world where `loader.js` is optional? (AKA: the more static we make things, the less we need `loader.js`)

The "namespace fallback" resolver slows down _all_ lookups, but if there are fewer resolutions required due to things being more static it might still be a tenable solution. It is also guaranteed to be the most terse of the solutions as the module names needn't be enumerated.

(Stream of consciousness response.)

**Comment by trentmwillis** [comment]
> Does this approach break `instanceOf` checks?

Unsure which specific checks you're referring to, but the reference returned by the module is the same. So objects will be the same as will classes, functions, etc.. This is the same as for traditional re-exports.

> What do we do in a world where `loader.js` is optional?

Is there actually a plan to drop `loader.js` any time in the near future?

> if there are fewer resolutions required due to things being more static it might still be a tenable solution.

Yes, local lookup + addon deduplication would be the best solution.

**Comment by rwjblue** [comment]
> However, the compiled code for this can be quite bloated:
> ```
> define('ember-chat/components/spanish-greeting', ['exports', 'common-components/components/spanish-greeting'], function (exports, _commonComponentsComponentsSpanishGreeting) {
>   'use strict';
> 
>   Object.defineProperty(exports, 'default', {
>     enumerable: true,
>     get: function get() {
>       return _commonComponentsComponentsSpanishGreeting['default'];
>     }
>   });
> });
```

This reminds me that we should be using `loose` mode (not an engines specific problem, but something we should likely address).

---

> Does this approach break instanceOf checks?

Not really.  In either case, if you import an exported `class` / factory from the reexport paths (or via `define.alias`) and compared the container instantiated instance via `instanceof` they would pass.

---

IMO, this isn't really an engine issue.  The import/reexport pattern that most addons do in `app/` can also receive the same treatment, but we need a way to ensure we are only using the alias treatment when no other code exists in the reexport.

**Comment by stefanpenner** [comment]
> Does this approach break instanceOf checks?

it should not (if it does, it is a bug) related tests -> https://github.com/ember-cli/loader.js/blob/master/tests/all.js#L743-L755

**Comment by stefanpenner** [comment]
> What do we do in a world where loader.js is optional? (AKA: the more static we make things, the less we need loader.js)

not worth thinking about

**Comment by nathanhammond** [comment]
> not worth thinking about, lets focus on problems of today.

I _strongly_ disagree with this sentiment. My thought was presented as, "hey, let's at least spend a few brain cycles on this to make sure we're happy with the solution even in a future state." This isn't an expensive exercise and forces us to analyze all of our constraints so we can come up with a best-of-breed solution. There was no intent to grandstand on my part by posing a question.

> Is there actually a plan to drop loader.js any time in the near future?

There's nothing to say that fully-static-resolution with module inlining isn't possible in some future state. In fact, we inch closer to that being a possibility the more work we do for modularization. Or, even in a transitional state, not all things would be guaranteed to be modules if they could be successfully inlined. That situation is more likely to be an issue.

---

I'm comfortable with either of a dynamic `resolver` or `define.alias` approach, we should simply be conscientious of the fact that we need to maintain a dep graph across the engine boundaries for `ember-modules-app` which would guarantee non-inlining so that it was addressable.

**Comment by stefanpenner** [comment]
@nathanhammond let me rephrase:

> What do we do in a world where loader.js is optional? (AKA: the more static we make things, the less we need loader.js)

After several brain cycles, I do not believe choices discussed here impact this anymore then the existing dynamism the system provides affects this.

As the feature is presented, it merely suggests a more concise form for an existing feature. Which itself, could make static analysis easier.

---

This led me to attempt to short-circuit, at the cost of clarity, I apologize for the brevity.

**Comment by trentmwillis** [comment]
I'm going to start exploring this as an AST transform as I feel it can likely have significant savings even when using loose mode for modules.

**Comment by trentmwillis** [comment]
A solution to this is being explored as a Babel plugin here: https://github.com/ember-engines/babel-plugin-compact-reexports
